go/ql/lib/semmle/go/internal/Locations.qll

```diff
@@ -0,0 +1,113 @@
+/** Provides classes for working with locations and program elements that have locations. */
+
+import go
+
+newtype TLocation =
+  TDbLocation(@location loc) or
+  TSynthLocation(string filepath, int startline, int startcolumn, int endline, int endcolumn) {
+    any(DataFlow::Node n).hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn) and
+    // avoid overlap with existing DB locations
+    not exists(File f |
+      locations_default(_, f, startline, startcolumn, endline, endcolumn) and
+      f.getAbsolutePath() = filepath
+    )
+  }
```

Comment:
We could have a single newtype branch, not sure if it is better, but I don't think being able to distinguish between extracted and synthetic locations is needed.
```suggestion
  TMkLocation(File file, int startline, int startcolumn, int endline, int endcolumn) {
    locations_default(_, file, startline, startcolumn, endline, endcolumn)
    or
    exists(DataFlow::Node n, string filepath |
      n.hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn) and
      file.getAbsolutePath() = filepath
    )
  }
```

Reply:
It is actually rather crucial to have them be separate injectors; otherwise evaluating `Locatable.getLocation` will require evaluating the data flow layer, which we want to avoid.

Reply:
Maybe there should be a comment explaining the rationale behind the way it is being done, to avoid someone changing it in future.

Reply:
I have updated the Go implementation to match that for JS: https://github.com/github/codeql/pull/15882.